src/struct_info.rs

```diff
@@ -279,6 +286,96 @@ impl<'a> StructInfo<'a> {
                     }
                 }
             }
+            #[doc(hidden)]
+            #[allow(dead_code, non_camel_case_types, non_snake_case)]
+            pub enum #repeated_fields_error_type_name {}
+            #[doc(hidden)]
+            #[allow(dead_code, non_camel_case_types, missing_docs)]
+            impl #impl_generics #builder_name < #( #target_generics ),* > #where_clause {
+                #[deprecated(
+                    note = #repeated_fields_error_message
+                )]
+                pub fn #field_name #generic_arg (self, _: #repeated_fields_error_type_name) -> #builder_name < #( #target_generics ),* > {
+                    self
+                }
+            }
+        })
+    }
+
+    pub fn required_field_impl(&self, field: &FieldInfo) -> Result<TokenStream, Error> {
+        let StructInfo {
+            ref name,
+            ref builder_name,
+            ..
+        } = *self;
+
+        let &FieldInfo {
+            name: ref field_name,
+            ..
+        } = field;
+        let mut builder_generics: Vec<syn::GenericArgument> = self
+            .generics
+            .params
+            .iter()
+            .map(|generic_param| match generic_param {
+                syn::GenericParam::Type(type_param) => {
+                    let ident = type_param.ident.clone();
+                    syn::parse(quote!(#ident).into()).unwrap()
+                }
+                syn::GenericParam::Lifetime(lifetime_def) => {
+                    syn::GenericArgument::Lifetime(lifetime_def.lifetime.clone())
+                }
+                syn::GenericParam::Const(const_param) => {
+                    let ident = const_param.ident.clone();
+                    syn::parse(quote!(#ident).into()).unwrap()
+                }
+            })
+            .collect();
+        let mut builder_generics_tuple = empty_type_tuple();
+        let generics = self.modify_generics(|g| {
+            for f in self.included_fields() {
+                if f.ordinal < field.ordinal && f.builder_attr.default.is_none() {
```

Comment:
This part is tricky to understand. Could you please add comments to each branch explaining why we do it? Something like:

```rust
if f.builder_attr.default.is_some() {
    // `f` is not mandatory - it does not have it's own fake `build` method, so `field` will need
    // to warn about missing `field` whether or not `f` is set.
    assert!(f.ordinal != field.ordinal, "`required_field_impl` called for optional field {}", field.name);
    g.params.push(f.generic_ty_param());
    builder_generics_tuple.elems.push_value(f.type_ident().into());
} else if f.ordinal < field.ordinal {
    // Only add a `build` method that warns about missing `field` if `f` is set. If `f` is not set,
    // `f`'s `build` method will warn, since it appears earlier in the argument list.
    builder_generics_tuple.elems.push_value(f.tuplized_type_ty_param());
} else if f.ordinal == field.ordinal {
    builder_generics_tuple.elems.push_value(empty_type());
} else {
    // `f` appears later in the argument list after `field`, so if they are both missing we will
    // show a warning for `field` and not for `f` - which means this warning should appear whether
    // or not `f` is set.
    g.params.push(f.generic_ty_param());
    builder_generics_tuple.elems.push_value(f.type_ident().into());
}
```

I've split the optional field case and the fields that come after case - it requires a bit of code duplication but I think it's more readable that way.